### DISCUSSION TITLE
Fully-qualified names with `nameof`

### DISCUSSION THREAD

**Discussion by Rabadash8820** [dotnet/csharplang#5483] [Language Ideas] [answered]
# Summary

I suggest that C# include a variant of `nameof` that generates an identifier's _fully-qualified_ name. I can think of two ways the language might allow this:
1. Providing an alternate operator. I considered `fullnameof` or `@nameof` (where the `@` evokes verbatim strings, which similarly modify the way _strings_ are "expanded" in some sense).
2. Providing an additional argument to the existing operator. Something like `nameof(Identifier, '.')`, where the second argument instructs the compiler to expand the identifier to its fully-qualified name, with the provided character (or any const string?) separating the parts. This has the benefit of generating a fully-qualified name with any delimiter without requiring "non-const" methods like `String.Split`/`String.Join`.

# Motivation

I often find myself wanting to map my namespace hierarchy to certain runtime hierarchies. For example, in the Unity game engine, I can define public methods for designers that appear nested within toolbar menus, and I typically like the nesting of the menus to match the nesting of my namespaces for easier maintenance. So if I have a method with fully qualified name `MyNamespace.SubNamespace.MyType.DoStuff`, I might nest it in the toolbar at "MyNamespace > SubNamespace > MyType > DoStuff". To instruct Unity to do this I declare my method as follows (Unity uses the `/` character to add a level of nesting:

```cs
[MenuItem($"{nameof(MyNamespace)}/{nameof(MyNamespace.SubNamespace)}/{nameof(MyType)}/{nameof(DoStuff)}")]
public static void DoStuff() => // ...
```

As you can see, I have to put a separate `nameof` operator around each part of the method's fully-qualified name. For namespaces, in particular, it would be nice if there were some way to get the fully-qualified name with one `nameof` call. Moreover, if I ever refactor things so that, e.g., `SubNamespace` becomes a top-level namespace, then all of a sudden my toolbar hierarchies will be out of sync, and there will be no compiler error/warning to tell me I'm identifying the namespace incorrectly.

Other scenarios where getting a fully-qualified type/member name might be useful include reflection, or really any scenario where someone doesn't want to "hard code" an identifier's namespaces.

**Comment by 333fred** [comment, marked as answer]
Duplicate of https://github.com/dotnet/csharplang/discussions/5200 and https://github.com/dotnet/csharplang/discussions/701.

**Reply by Rabadash8820** [reply]
Damn...I promise I searched the Issues page before creating this 😅, guess I should've checked the Discussions page...

FWIW, I still think the secord-argument syntax is still worth considering. Guess I'll mention that on #701.

**Comment by jnm2** [comment]
I'm proposing making `typeof(`...`).FullName` a constant, which will help with some of these use cases: https://github.com/dotnet/csharplang/issues/8505

It would be great to hear in that thread about real-world use cases that support doing this, as well as real-world use cases for `fullnameof` that _aren't_ addressed by that proposal and don't have a good solution today.
